src/analysis/special_functions/trigonometric.lean

```diff
@@ -2287,6 +2359,19 @@ end chebyshev₁
 namespace real
 open_locale real
 
+lemma tan_add {x y : ℝ}
+  (h : ((∀ k : ℤ, x ≠ (2 * k + 1) * π / 2) ∧ ∀ l : ℤ, y ≠ (2 * l + 1) * π / 2)
+     ∨ ((∃ k : ℤ, x = (2 * k + 1) * π / 2) ∧ ∃ l : ℤ, y = (2 * l + 1) * π / 2)) :
+  tan (x + y) = (tan x + tan y) / (1 - tan x * tan y) :=
+by simpa only [← complex.of_real_inj, complex.of_real_sub, complex.of_real_add, complex.of_real_div,
+              complex.of_real_mul, complex.of_real_tan]
+    using @complex.tan_add (x:ℂ) (y:ℂ) (by { convert h; norm_cast })
```

Comment:
Here you could have written
```lean
begin
  have h' : ((∀ k : ℤ, (x : ℂ) ≠ (2 * k + 1) * π / 2) ∧ ∀ l : ℤ, (y : ℂ) ≠ (2 * l + 1) * π / 2)
     ∨ ((∃ k : ℤ, (x : ℂ) = (2 * k + 1) * π / 2) ∧ ∃ l : ℤ, (y : ℂ) = (2 * l + 1) * π / 2),
  exact_mod_cast h,
  exact_mod_cast (complex.tan_add h')
end
```
to save you pain, but again this is very slow.

Reply:
Is `exact_mod_cast` significantly more expensive than `norm_cast`? @robertylewis

Reply:
`exact_mod_cast h` is equivalent to `norm_cast, norm_cast at h, exact h`.

Reply:
So normally (assuming `exact` is inexpensive) they should have the same cost?

Reply:
It very much depends on what you're calling them on. `exact_mod_cast` calls `norm_cast` twice so it "should" have twice the cost, but often one side is already simplified or something, so it's cheaper.